sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala

```diff
@@ -0,0 +1,211 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.streaming
+
+import java.lang.Thread.UncaughtExceptionHandler
+
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.spark.Logging
+import org.apache.spark.sql.{ContinuousQuery, DataFrame, SQLContext}
+import org.apache.spark.sql.catalyst.expressions.{Attribute, AttributeMap}
+import org.apache.spark.sql.catalyst.plans.logical.{LocalRelation, LogicalPlan}
+import org.apache.spark.sql.catalyst.util._
+import org.apache.spark.sql.execution.QueryExecution
+
+/**
+ * Manages the execution of a streaming Spark SQL query that is occurring in a separate thread.
+ * Unlike a standard query, a streaming query executes repeatedly each time new data arrives at any
+ * [[Source]] present in the query plan. Whenever new data arrives, a [[QueryExecution]] is created
+ * and the results are committed transactionally to the given [[Sink]].
+ */
+class StreamExecution(
+    sqlContext: SQLContext,
+    private[sql] val logicalPlan: LogicalPlan,
+    val sink: Sink) extends ContinuousQuery with Logging {
+
+  /** An monitor used to wait/notify when batches complete. */
+  private val awaitBatchLock = new Object
+
+  @volatile
+  private var batchRun = false
+
+  /** Minimum amount of time in between the start of each batch. */
+  private val minBatchTime = 10
+
+  /** Tracks how much data we have processed from each input source. */
+  private[sql] val streamProgress = new StreamProgress
+
+  /** All stream sources present the query plan. */
+  private val sources =
+    logicalPlan.collect { case s: StreamingRelation => s.source }
+
+  // Start the execution at the current offsets stored in the sink. (i.e. avoid reprocessing data
+  // that we have already processed).
+  {
+    sink.currentOffset match {
+      case Some(c: CompositeOffset) =>
+        val storedProgress = c.offsets
+        val sources = logicalPlan collect {
+          case StreamingRelation(source, _) => source
+        }
+
+        assert(sources.size == storedProgress.size)
+        sources.zip(storedProgress).foreach { case (source, offset) =>
+          offset.foreach(streamProgress.update(source, _))
+        }
+      case None => // We are starting this stream for the first time.
+      case _ => throw new IllegalArgumentException("Expected composite offset from sink")
+    }
+  }
+
+  logInfo(s"Stream running at $streamProgress")
+
+  /** When false, signals to the microBatchThread that it should stop running. */
+  @volatile private var shouldRun = true
+
+  /** The thread that runs the micro-batches of this stream. */
+  private[sql] val microBatchThread = new Thread("stream execution thread") {
+    override def run(): Unit = {
+      SQLContext.setActive(sqlContext)
+      while (shouldRun) {
+        attemptBatch()
```

Comment:
There is already a uncaughtexception handler for this thread.

On Tue, Feb 2, 2016 at 3:29 PM, tedyu <notifications@github.com> wrote:

> In
> sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala
> <https://github.com/apache/spark/pull/11006#discussion_r51654640>:
>
> > +      case None => // We are starting this stream for the first time.
> > +      case _ => throw new IllegalArgumentException("Expected composite offset from sink")
> > +    }
> > +  }
> > +
> > +  logInfo(s"Stream running at $streamProgress")
> > +
> > +  /** When false, signals to the microBatchThread that it should stop running. */
> > +  @volatile private var shouldRun = true
> > +
> > +  /** The thread that runs the micro-batches of this stream. */
> > +  private[sql] val microBatchThread = new Thread("stream execution thread") {
> > +    override def run(): Unit = {
> > +      SQLContext.setActive(sqlContext)
> > +      while (shouldRun) {
> > +        attemptBatch()
>
> Should this be wrapped in try block ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/11006/files#r51654640>.
>